Cython/Utility/ObjectHandling.c

```diff
@@ -2553,3 +2553,132 @@ static PyObject *__Pyx_PyMethod_New(PyObject *func, PyObject *self, CYTHON_UNUSE
+    #else
+        #define __PYX_PYUNICODE_CONCATINPLACE(a, b) __Pyx_PyUnicode_ConcatInPlaceImpl(&a, b, NULL)
+    #endif
+// __PYX_PYUNICODE_CONCATINPLACE is slightly odd because it has the potential to modify the input
+// argument (but only in cases where no user should notice). Therefore, it needs to keep Cython's
+// refnanny informed.
+static CYTHON_INLINE PyObject *__Pyx_PyUnicode_ConcatInPlaceImpl(PyObject **a, PyObject *b, void * __pyx_refnanny); /* proto */
+#else
+    #define __PYX_PYUNICODE_CONCATINPLACE __Pyx_PyUnicode_Concat
+#endif
+#define __PYX_PYUNICODE_CONCATINPLACESafe(a, b) ((unlikely((a) == Py_None) || unlikely((b) == Py_None)) ? \
+    PyNumber_Add(a, b) : __PYX_PYUNICODE_CONCATINPLACE(a, b))
+
+////////////// UnicodeConcatInPlace ////////////////////
+
+#if CYTHON_COMPILING_IN_CPYTHON && PY_MAJOR_VERSION > 2
+    static CYTHON_INLINE PyObject *__Pyx_PyUnicode_ConcatInPlaceImpl(PyObject **a, PyObject *b, void * __pyx_refnanny) {
+    __Pyx_GIVEREF(*a);
+    PyUnicode_Append(a, b);  // this does it's own safety checks so is always safe to use
+    __Pyx_XINCREF(*a);  // there are two copies of `*a` - the one just returned and the one
+                // passed in as an argument
+    return *a;
+}
+#endif
+
+//////////// BytesConcat.proto ///////////////////////
+
+// Follows the interface of "PyUnicode_Concat", but with Bytes, rather than the
+// interface of PyBytes_Concat. However, uses PyBytes_Concat which has the potential
+// to modify in-place. See UnicodeConcatInPlace for comments
+#if CYTHON_COMPILING_IN_CPYTHON
+    #if CYTHON_REFNANNY
+        #define __PYX_PYBYTES_CONCATINPLACE(a, b) __Pyx_PyBytes_ConcatInPlaceImpl(&a, b, __pyx_refnanny)
+    #else
+        #define __PYX_PYBYTES_CONCATINPLACE(a, b) __Pyx_PyBytes_ConcatInPlaceImpl(&a, b, NULL)
+    #endif
+    static CYTHON_INLINE PyObject *__Pyx_PyBytes_ConcatInPlaceImpl(PyObject **a, PyObject *b, void * __pyx_refnanny); /* proto */
+    static CYTHON_INLINE PyObject *__Pyx_PyBytes_Concat(PyObject *a, PyObject *b);
+#else
+    #define __Pyx_PyBytes_Concat(a, b) PyNumber_Add(a,b)
+    #define __PYX_PYBYTES_CONCATINPLACE(a, b) PyNumber_Add(a,b)
+#endif
+#define __Pyx_PyBytes_ConcatSafe(a, b) ((unlikely((a) == Py_None) || unlikely((b) == Py_None)) ? \
+    PyNumber_Add(a, b) : __Pyx_PyBytes_Concat(a, b))
+#define __PYX_PYBYTES_CONCATINPLACESafe(a, b) ((unlikely((a) == Py_None) || unlikely((b) == Py_None)) ? \
+    PyNumber_Add(a, b) : __PYX_PYBYTES_CONCATINPLACE(a, b))
+
+//////////// BytesConcat ///////////////////////
+
+#if CYTHON_COMPILING_IN_CPYTHON
+static CYTHON_INLINE PyObject *__Pyx_PyBytes_ConcatInPlaceImpl(PyObject **a, PyObject *b, void * __pyx_refnanny) {
+   __Pyx_GIVEREF(*a);
+    PyBytes_Concat(a, b);  // this does it's own safety checks so is always safe to use
+    __Pyx_XINCREF(*a);  // there are two copies of `*a` - the one just returned and the one
+                // passed in as an argument
+    return *a;
+}
```

Comment:
This looks like it should just be a C macro instead of an inline function. That would avoid the need for passing the refnanny context.

Reply:
I can't make a macro work unfortunately. It needs to return a value and using the comma operator fails because `__Pyx_GIVEREF` and `__Pyx_INCREF` are macros themselves that expand to `do {...} while(0)`